src/contract.ts

```diff
@@ -0,0 +1,279 @@
+    public get connection(): Connection {
+        return this._connection;
+    }
+
+    private _contractId: string;
+    public get contractId(): string {
+        return this._contractId;
+    }
+
+    private _abi: ABI;
+    public get abi(): ABI {
+        return this._abi;
+    }
+
+    /**
+     * @param connection Connection to NEAR network through RPC.
+     * @param contractId NEAR account id where the contract is deployed.
+     * @param abi ABI schema which will be used to generate methods to be called on this Contract
+     */
+    constructor(connection: Connection, contractId: string, abi: ABI) {
+        this._connection = connection;
+        this._contractId = contractId;
+        this._abi = abi;
+
+        this._abi.abi.functions.forEach((fn) => {
+            const funcName = fn.name;
+            const isView = fn.is_view;
+            // Create method on this contract object to be able to call methods.
+            Object.defineProperty(this, funcName, {
+                writable: false,
+                enumerable: true,
+                value: (...args: any[]) => {
+                    const connection = this._connection;
+                    const contractId = this._contractId;
+                    // TODO support providing an object as a single parameter
+                    if (fn.params && args.length !== fn.params.length) {
+                        throw new AbiValidationError(
+                            `Invalid parameter length for method ${fn.name}, expected ${fn.params.length}`
+                        );
+                    }
+                    const func: CallableFunction = {
+                        // Include a call function if the function is not view only.
+                        callFrom: isView
+                            ? undefined
+                            : async function (
+                                account,
+                                opts
+                            ): Promise<FinalExecutionOutcome> {
+                                // Using inner NAJ APIs for result for consistency, but this might
+                                // not be ideal API.
+                                return callInternal(
+                                    account,
+                                    contractId,
+                                    funcName,
+                                    serializeArgs(funcName, args, fn.params),
+                                    opts
+                                );
+                            },
+                        view: async function (): Promise<any> {
+                            const returnBytes = await viewInternal(
+                                connection,
+                                contractId,
+                                funcName,
+                                serializeArgs(funcName, args, fn.params)
+                            );
+                            // TODO deserialize based on protocol from schema
```

Comment:
I don't think you can deserialize based on a schema, but you can definitely validate whether a given JSON conforms to the JSON Schema which could potentially be done here (and the same thing with user-supplied arguments).

Reply:
Well, right now we are just deserializing from a buffer into the JS object, but what I'm referring to is being able to deserialize as borsh instead of JSON if it's included in the schema, for example. Validation of the returned object could be valuable too though

Reply:
Ah, yeah forgot about borsh